runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1651,6 +1707,277 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
+      if (trace())
+         {
+         traceMsg(comp(), "%s %s fieldSymRef: %s fieldEntry[%d] fieldName %s fieldSig %s type %d offset %d\n", __FUNCTION__,
+            comp()->getDebug()->getName(loadOpCode), comp()->getDebug()->getName(fieldSymRef), (int)idx,
+            fieldEntry._fieldname, fieldEntry._typeSignature, fieldEntry._datatype.getDataType(), fieldEntry._offset);
+         }
+
+      TR::Node *loadNode = TR::Node::createWithSymRef(loadOpCode, 1, elementAddressNode, 0, fieldSymRef);
+      newValueNode->setAndIncChild(newValueNodeChildIndex++, loadNode);
+      }
+   }
+
+bool
+J9::ValuePropagation::transformFlattenedArrayElementStore(TR_OpaqueClassBlock *arrayClass, TR::TreeTop *callTree, TR::Node *callNode)
+   {
+   /*
+    * primitive public class MultipleFieldsPrimitive {
+    *    public final int x;
+    *    public final int y;
+    * }
+    * primitive public class NestedFieldsPrimitive {
+    *    public final MultipleFieldsPrimitive vtFiled0;
+    *    public final MultipleFieldsPrimitive vtFiled1;
+    * }
+    *
+    * Before:
+    * n9n       treetop
+    * n8n         call  jitStoreFlattenableArrayElement
+    * n5n           aload  <parm 2 QNestedFieldsPrimitive;>
+    * n4n           iload  <parm 0 I>
+    * n3n           aload  <parm 1 [QNestedFieldsPrimitive;>
+    *
+    * After:
+    * n9n       treetop
+    * n8n         istorei  <Q-NestedFieldsPrimitive-array-shadow>.vtFiled0.x
+    * n3123n        aladd
+    * n3n             aload  <parm 1 [QNestedFieldsPrimitive;>
+    * n3122n          ladd
+    * n3120n            lshl
+    * n3118n              i2l
+    * n4n                   iload  <parm 0 I>
+    * n3119n              iconst 4
+    * n3121n            lconst 16
+    * n3124n        iloadi  NestedFieldsPrimitive.vtFiled0.x
+    * n5n             aload  <parm 2 QNestedFieldsPrimitive;>
+    * n3156n    istorei  <Q-NestedFieldsPrimitive-array-shadow>.vtFiled0.y
+    * n3153n      ==>aladd
+    * n3155n      iloadi  NestedFieldsPrimitive.vtFiled0.y
+    * n5n           ==>aload
+    * n3158n    istorei  <Q-NestedFieldsPrimitive-array-shadow>.vtFiled1.x
+    * n3153n      ==>aladd
+    * n3157n      iloadi  NestedFieldsPrimitive.vtFiled1.x
+    * n5n           ==>aload
+    * n3160n    istorei  <Q-NestedFieldsPrimitive-array-shadow>.vtFiled1.y
+    * n3153n      ==>aladd
+    * n3159n      iloadi  NestedFieldsPrimitive.vtFiled1.y
+    * n5n           ==>aload
+    *
+    */
+   TR_OpaqueClassBlock *arrayComponentClass = fe()->getComponentClassFromArrayClass(arrayClass);
+
+   const TR::TypeLayout *fieldTypeLayout = comp()->typeLayout(arrayComponentClass);
+   size_t fieldCount = fieldTypeLayout->count();
+
+   TR::Node *valueNode = callNode->getChild(0);
+   TR::Node *indexNode = callNode->getChild(1);
+   TR::Node *arrayRefNode = callNode->getChild(2);
+
+   // The value that is being stored into the array element has to be non null.
+   if (!valueNode->isNonNull())
```

Comment:
Also note that Vijay had pointed out in a review of one of my value propagation pull requests that the check of non-null ought to look at the constraint rather than the node - though I have to admit that I didn't check at what point `isNull` and `isNonNull` actually get set on the nodes themselves.

Reply:
Thanks for the suggestion! I'll split `RequiresStoreAndNullCheck` into `RequiresStoreCheck` and `RequiresNullValueCheck`, so that `RequiresNullValueCheck` can be reused in the unflattened and flattened cases.

Reply:
Addressed in e81ca9b